[Logs Explorer][Discover] Move Logs Overview into Discover codebase

tonyghiani opened the pull request:
## 📓 Summary

Closes #180024 

This work aims to move the "Logs Overview" doc view created from the logs-explorer app into the unified-doc-viewer plugin, creating and registering this as a preset along the table and source doc views.

To keep control of whether this doc view should be displayed or not, an `enabled` configuration flag is supported for every doc view and will be used to determine whether a doc view should load or not in the view.
This `enabled` flag can be programmatically enabled/disabled by `docView.id` using the 2 new methods added to the `DocViewsRegistry`, the `enableById` and `disableById` ones.
The customization extension point does not register the content of the tab, but is limited to enable/disable a preset overview tab.

To allow this change, some shared utilities between the flyout logic and the smart fields feature have been copied into the `@kbn/discover-utils` package. The utils will still live in the logs_explorer plugin and are used by the smart fields feature until this is migrated too into Discover.

## 💡 Reviewer hints

Although it seems a large PR, most of the changes are on moved files from logs-explorer into unified-doc-viewer, which is logic already reviewed. With these changes, there will be a follow-up PR to optimize the shared parts with the other doc views.

## 🚶 Next steps

To keep the scope of this PR the smallest possible, here are some changes I'll work out in upcoming PRs built on top of this one.

- [x] Implement a discover registry to enable registering external features, such as the ObservabilityAIAssistant. 
- [x] Integrate ObsAIAssistant with this work through the new discover features registry.
- [x] Refactor the doc views to share duplicated logic.
- [x] Port the existing e2e tests for the logs overview tab into the unified-doc-viewer plugin.

https://github.com/tonyghiani/kibana/pull/1

apmmachine commented:
### :robot: GitHub comments

<details><summary>Expand to view the GitHub comments</summary>
<p>

Just comment with:
- `/oblt-deploy` : Deploy a Kibana instance using the Observability test environments.
- `/oblt-deploy-serverless` : Deploy a serverless Kibana instance using the Observability test environments.
- `run` `elasticsearch-ci/docs` : Re-trigger the docs validation. (use unformatted text in the comment!)

</p>
</details>

tonyghiani commented:
/ci

tonyghiani commented:
/ci

tonyghiani commented:
/ci

tonyghiani commented:
/ci

tonyghiani commented:
The reason behind the huge move of async bundle chunks from logs explorer into unified doc viewer is due to the ecs package that loads the fields descriptions.

elasticmachine commented:
Pinging @elastic/kibana-data-discovery (Team:DataDiscovery)

elasticmachine commented:
Pinging @elastic/obs-ux-logs-team (Team:obs-ux-logs)

flash1293 commented:
Thanks for putting this together @tonyghiani ! I think the "enable" approach is well suited for now.

When introducing the new architecture with backbone and flavor layer in place, I wonder whether that even needs to be dynamic or whether it could just be something that's defined in a fixed way in the flavor layer - this would further reduce the moving parts and make it easier to understand why a certain thing is rendered in a certain situation.

tonyghiani commented:
Thanks for the review @jughosta !

> Would be nice to create a separate package for Logs Overview tab and move most of the related utils from `discover-utils` and components from `unified-doc-viewer` to it.

I added them into the unified-doc-viewer to keep the doc views preset well organized together, splitting it into a different package IMO feels a bit overengineered since it represents a doc view that can only be (conditionally) used along the table and source doc views.

> This way we can assign ownership for such package to your team.

If ownership is the main concern we can apply a more specific rule for this doc view folder and assign it to the logs team, let me know if this sounds good and I'll make the change.

Regarding the utils/types added to `discover-utils`, I moved them there since they share logic used for the smart fields feature, which is most likely the next refactor we'll take on, so having those in a utils package dedicated to discover felt right.

Let me know your thought, I'm happy to bring in any change you feel would make this part clearer to our teams.

weltenwort reviewed (commented):
Nice job, only had a few nits to pick below :innocent: :+1: 

I was wondering if it would be worth keeping at least a basic functional test for the fly-out that checks whether the overview tab is there. What do you think?

tonyghiani commented on packages/kbn-discover-utils/src/field_constants.ts:
Sure, it was a leftover as some constants are still used by the smart fields implementation, I'll add a re-export from the package. Also, the file holding the logs explorer constants will be more likely deleted or reduced after all the features refactor into discover work.

tonyghiani commented on src/plugins/unified_doc_viewer/public/hooks/use_field_actions.tsx:
```diff
@@ -0,0 +1,131 @@
+      {
+        id: 'addFilterInAction',
+        iconType: 'plusInCircle',
+        label: actionFilterForText(value as string),
```
The suggestion makes sense, and double-checking what is used for the table actions we don't even need to contact the value to the tooltip, as it would be different from the table actions. I will remove the need to contact the value on the tooltip (which is just shown next to the action) so that it also makes it easier to re-use these actions for the available doc views (upcoming refactoring on this).

tonyghiani commented on src/plugins/unified_doc_viewer/public/components/doc_viewer_logs_overview/logs_overview.tsx:
```diff
@@ -0,0 +1,41 @@
+  onRemoveColumn,
+}: DocViewRenderProps) {
+  const { fieldFormats } = getUnifiedDocViewerServices();
+  const parsedDoc = getDocumentOverview(hit, { dataView, fieldFormats });
```
I didn't go for memoization here since for most of the interactions that could cause a rerender such as filtering for a field, the `hit` property would change (contains `highlights`) and consequently nullify the memoization anyway, which seemed a de-optimization.

As a side note, I triggered multiple rerenders with 6x CPU throttling, and the function execution is always in a 0.1-0.3ms range.

tonyghiani commented on src/plugins/unified_doc_viewer/public/components/doc_viewer_logs_overview/logs_overview_highlights.tsx:
```diff
@@ -0,0 +1,315 @@
+  );
+}
+
+const serviceLabel = i18n.translate('unifiedDocViewer.docView.logsOverview.label.service', {
```
I think this comment is outdated but I remember the capitalization of the labels, you are right about the reason, it was due to bulk renaming operations 😇

weltenwort replied:
oops, yes... apparently I was looking at an older revision, sorry

tonyghiani commented on packages/kbn-unified-doc-viewer/src/services/doc_views_registry.test.tsx:
```diff
@@ -74,6 +74,25 @@ describe('DocViewerRegistry', () => {
   });
 
+  describe('#enableById & #disableById', () => {
+    test('should enable/disable a doc view given the passed id', () => {
```
Agree, I'll add a case for this in the clone test, thanks for the heads-up!

tonyghiani commented on packages/kbn-unified-doc-viewer/src/services/doc_views_registry.ts:
```diff
@@ -50,8 +56,30 @@ export class DocViewsRegistry {
     this.docViews.delete(id);
   }
 
+  enableById(id: string) {
```
Agree, ideally we could initialize a DocViewsRegistry instance for a specific flavoured version of discover and set its doc views, this would remove the need for cloning the registry as every version would have its own instance, or at list the control of the registries would be centralized in discover. We can definitely tackle this once the code-base is more unified, meanwhile, I improved the test of clone to assert against mutations.

tonyghiani commented:
@weltenwort thanks for the review! I addressed all your suggestions, let me know if you feel something else should change.

> I was wondering if it would be worth keeping at least a basic functional test for the fly-out that checks whether the overview tab is there. What do you think?

Agree, I restored the initial `flyout.ts` test files and left a short assertion against the tab existence.

jughosta reviewed (approved):
LGTM

weltenwort reviewed (commented):
Thanks for creating the smoke tests. And if it works out, I like the use of the flex grid!

I'm somehow struggling to deploy this to a cloud instance, so I wasn't able to check it visually yet.

I've seen several places where you removed intermediate `<EuiFlexItem>`s inside of `<EuiFlexGroup>`s. Was that intentional? Why do we need the groups in the first place if they don't contain flex items?

davismcphee commented on src/plugins/unified_doc_viewer/public/components/doc_viewer_logs_overview/sub_components/highlight_field_description.tsx:
```diff
@@ -17,12 +18,8 @@ export function HighlightFieldDescription({ fieldName }: { fieldName: string })
-        </EuiFlexItem>
-      )}
-      <EuiFlexItem grow={false}>{fieldName}</EuiFlexItem>
+      {type && <FieldIcon type={type} size="s" />}
+      {fieldName}
```
@weltenwort mentioned this already, but I also wouldn't expect to see `EuiFlexGroup` used without `EuiFlexItem`. Was there a reason these changes were made?

tonyghiani replied:
I explained more in detail on [this comment](https://github.com/elastic/kibana/pull/180262#issuecomment-2074183685).

davismcphee commented on x-pack/test/functional/apps/observability_logs_explorer/flyout.ts:
```diff
@@ -65,32 +65,9 @@ export default function ({ getService, getPageObjects }: FtrProviderContext) {
-      await testSubjects.existOrFail('logsExplorerFlyoutDetail');
-    });
-
-    it('should display a timestamp badge', async () => {
```
Should these tests have been removed? I see this task is completed in the PR summary, but they don't seem to have been ported anywhere:
>Port the existing e2e tests for the logs overview tab into the unified-doc-viewer plugin.

Personally I'd be fine with leaving hem here with updated `data-test-subj` values until we can test them in Discover once enabled.

tonyghiani replied:
These tests have been moved and adapted next to the doc view implementation: https://github.com/elastic/kibana/pull/180262/files#diff-35dde01753f22a9d917cb933594571e1d9583419455c0a00223e313bd2dd32a6

davismcphee reviewed (commented):
The code changes look good and it worked well when testing! I left a few questions, but otherwise this looks good on my end.

tonyghiani commented:
cc. @weltenwort

> I've seen several places where you removed intermediate <EuiFlexItem>s inside of <EuiFlexGroup>s. Was that intentional? 

I recently had an interesting discussion with the EUI team on an issue I created, and an [off-topic comment](https://github.com/elastic/eui/issues/7612#issuecomment-2024085689) (the first half of the comment is what matters) directly touched on the usage of EuiFlexGroup + EuiFlexItem.

The EuiFlexGroup + EuiFlexItem combo was in place to achieve the `gap` behaviour on flex layouts, which eliminated the requirement of using both in combination once the team switched to the native `gap` CSS property.
I agree with them it doesn't add benefits except when we need to control the content stretch of a flex layout, they also recommended removing extra div wrappers whether is possible.

> Why do we need the groups in the first place if they don't contain flex items?

`EuiFlexGroup` might now be a bit misleading as the component name, as it's only applying CSS flex properties to a div element. The places where I left it are mostly for aligning and spacing between items.

tonyghiani commented on packages/kbn-discover-utils/src/utils/get_document_overview.ts:
```diff
@@ -0,0 +1,89 @@
+import type { DataView } from '@kbn/data-views-plugin/public';
+import { DataTableRecord, DocumentOverview, fieldConstants, formatFieldValue } from '../..';
+
+export function getDocumentOverview(
```
Addressed in https://github.com/elastic/kibana/pull/180262/commits/88b678714daf5b8c29294d497bee74f3568d0872

tonyghiani commented on packages/kbn-discover-utils/src/types.ts:
```diff
@@ -46,3 +46,41 @@ type FormattedHitPair = readonly [
  */
 export type FormattedHit = FormattedHitPair[];
+
+export interface DocumentOverview extends ResourceFields, StackTraceFields, CloudFields {
```
Addressed in https://github.com/elastic/kibana/pull/180262/commits/d2647a69466637ca50c126fd45aa894db7111a7e

weltenwort commented:
Thanks for teaching me about that new flexbox gap semantic. It makes sense to avoid unnecessary `divs`.

weltenwort reviewed (approved):
I really like the thought of relying on flex grid to perform the column calculations for us :+1: 

During testing I found that it behaves differently from the previous implementation:

- It switches to two columns earlier when the screen is small
- It switches to more than three columns when the screen is large, while three was the upper limit previously

![image](https://github.com/elastic/kibana/assets/973741/c9b521ce-be75-47f8-a84d-2cd2f5fde3b5)

I think the new behavior doesn't feel any worse than the previous one. Just wanted to point this out in case there was an important reason for the previous behavior. Maybe @isaclfreire can double-check?

weltenwort commented on x-pack/plugins/observability_solution/logs_explorer/public/customizations/custom_flyout_content.tsx:
```diff
@@ -55,7 +54,8 @@ const CustomFlyoutContent = ({
   <EuiFlexItem>
-    <FlyoutDetail actions={actions} dataView={dataView} doc={doc} />
+    {/* TOREMOVE */}
+    {/* <FlyoutDetail actions={actions} dataView={dataView} doc={doc} /> */}
```
Did you keep this intentionally?

tonyghiani replied:
I did, it was to keep a reference on this touchpoint as I planned (and did in https://github.com/tonyghiani/kibana/pull/1/files#diff-3933fc8ad52c2ec68c912e7099f3ab7b5a291652eac29dd1271a6dc43df3223b) to remove the whole file.

tonyghiani commented:
@weltenwort @isaclfreire I agree with Felix as it makes easier to tweak and maintain the columns layout using native CSS.
On the details, we can also use the `auto-fill` approach to keep columns aligned where there are less values filling the whole space:

**auto-fill**

<img width="1125" alt="Screenshot 2024-04-25 at 11 00 58" src="https://github.com/elastic/kibana/assets/34506779/c0f22c49-688a-43d9-8757-b31bb69daeed">

**auto-fit**

![image](https://private-user-images.githubusercontent.com/973741/325285144-c9b521ce-be75-47f8-a84d-2cd2f5fde3b5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQwMzUwMDMsIm5iZiI6MTcxNDAzNDcwMywicGF0aCI6Ii85NzM3NDEvMzI1Mjg1MTQ0LWM5YjUyMWNlLWJlNzUtNDdmOC1hODRkLTJjZDJmNWZkZTNiNS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNDI1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDQyNVQwODQ1MDNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kYWVlMDg2NGJiMGI5MTVlNTdlNjdkMzgwMmNkZjIzOTRiNWVhZGEyYTU2ZTFmYjNiYTdkNTNkMmUzMWU1YWJjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.IEFbY4P7Xm3VvNmQDTZESxgGvsQcAlqo0V0iR-r9JqA)

@isaclfreire do you have any preference on the above options?
